### ISSUE TITLE
Problem with external JavaScript

### ISSUE THREAD

**Issue by IanStorm** [AngleSharp/AngleSharp.Scripting#45]
Hello,

I'm trying to parse information from a website whose content creation relies on JavaScript; on external JavaScript files, to be precise. Somehow AngleSharp doesn't seem to work in that particular scenario.
I was able to create a MWE of my issue:

.html:
```html
<!doctype html>
<html>
    <head><title>Sample</title></head>
    <body>
        <script src="test.js" type="text/javascript"></script>
    </body>
</html>
```

test.js:
```javascript
document.title = 'Simple manipulation...';
document.write('<span class=greeting>Hello World!</span>');
```

The AngleSharp `HtmlParser` that I am using is generated as follows:
```csharp
var config = Configuration.Default
    .WithDefaultLoader(setup => setup.IsResourceLoadingEnabled = true)
    .WithJavaScript();
var parser = new HtmlParser(config);
```

As you can see, the example is pretty close to the one from the [AngleSharp wiki](https://github.com/AngleSharp/AngleSharp/wiki/Examples#connecting-javascript-evaluation).

I already saw some issues that seem to be related (https://github.com/AngleSharp/AngleSharp.Scripting/issues/35, https://github.com/AngleSharp/AngleSharp.Scripting/issues/44, https://github.com/AngleSharp/AngleSharp.Scripting/issues/43, https://github.com/AngleSharp/AngleSharp.Scripting/issues/24). Still, I can't quite figure out a solution to my problem.
Can somebody please help?
Thanks in advance.

**Comment by FlorianRappl** [comment]
Any reason why you don't use `BrowsingContext`?

That being written - I do not see your problem ("does not work", i.e., what is the result you expect?). I guess you use (or should use) `ParseAsync`. The script is - in any way - executed asynchronously. Thus are you sure that you do not get any result at any time?

**Comment by IanStorm** [comment]
Thank you for that quick response.

> That being written - I do not see your problem ("does not work", i.e., what is the result you expect?).

Good point. I should have mentioned this in a more clear way.
I am expecting the following HTML as a result of the parsing:
```html
<!doctype html>
<html>
    <head><title>Sample</title></head>
    <body>
        <script src="test.js" type="text/javascript"></script>
        <span class=greeting>Hello World!</span>
    </body>
</html>
```
Meaning that I am expecting a `span` tag inside the body.
But I am receiving the input HTML as output, i.e., without any `span` tag.

> I guess you use (or should use) `ParseAsync`. The script is - in any way - executed asynchronously.

I am using the `ParseAsync` method. I also tried with `Parse`. But if it's executed asynchronously in any way, that at least makes it clear why `Parse` did not work either.

> Thus are you sure that you do not get any result at any time?

Yes and no. Maybe _sure_ is the wrong word:
"Yes", because I tried checking the content, i.e., `Body.InnerHtml`, of the returned document, then doing `Thread.Sleep(1000)`, and checking it again. I even did this 10 times in a row, meaning a total wait time of 10 seconds.
"No", because there's maybe a misunderstanding and I am doing it wrong.

> Any reason why you don't use `BrowsingContext`?

There's not been a specific reason for that, no. It was basically because the JavaScript example from your wiki was using the `HtmlParser`, and I simply started by copying the example.
However, I just tried using the `BrowsingContext`; the result is the same. Even if I wait for some time, the document remains unchanged.

To summarize:
Independent from using `BrowsingContext` or `HmtlParser`, the `span` tag is not added to the body of the returned document. This behavior is also independent from any wait time.
Finally, just to make this clear, if I add the content of the `test.js` file to the HTML, everything works fine. It only fails in case of referencing.

**Comment by FlorianRappl** [comment]
Hi @IanStorm - thanks for the clarification.

I think everything is working alright - the major issue seems to be the resolution of test.js. It is resolved relative to the HTML path, which (as you use the HtmlParser directly) is set to about:blank (no URL given). If you use the `BrowsingContext` you can use the virtual response in order to set a proper path for your input document.

```cs
context.openAsync((res) => res.Content(...).Address(...))
```

**Comment by IanStorm** [comment]
Right after writing my response, I suddenly realized about the possible issue: I said...

> [...] and I simply started by copying the example.

I first tried using the exact code, and then extended it to more and more reflect my scenario; to see where _exactly_ if starts failing.
The thing that I am doing, is to first convert the HTML into a string and pass that as a parameter to the AngleSharp parser. That being said, how should the parser know about the external resource? As it's only receiving a string, it is not able to resolve the relative URL given in the HTML.
I guess that's the silly thing that I've done wrong while slowly adapting the example to the target scenario.

That means, I will now verify whether this is correct, and update this issue later. I guess/hope that no further action is required from your side.

**Comment by IanStorm** [comment]
Oh, while writing my response, I did not see your comment, @FlorianRappl .

Yeah, your input is matching to what I've written.
Thank you. And sorry for bothering; it's a silly mistake from my side.

As I've said before, I will check whether this is correct and then update you here later.

**Comment by FlorianRappl** [comment]
Alright, thanks for the info @IanStorm !

**Comment by IanStorm** [comment]
@FlorianRappl ,
as discussed, this is my "callback".

I can solve the issue with the external JavaScript file using your proposed solution.
So, thank you for this. 👍 

However, another issue opens up by that (if you prefer, we can close this issue and I can open another):
This approach only works for files that can be accessed via `http://…`; not for ones located on my local disk, i.e., files that are accessed via `file:///…`. It's not a problem for the production scenario, as I am only accessing files via HTTP there. But in case of unit testing it seems more convenient to me to provide access to files stored on the local machine (and therefore via `file:///…`).
Can you comment on this? Is this intended / by design? Do you know any other way than simulating HTTP access for the tests also?

**Comment by FlorianRappl** [comment]
Hm I don't know if this was built into v0.9.9 of AngleSharp. I don't think so. But AngleSharp.Io also brings a requester that can be used with the file:// scheme. Or you can roll your own of course (just need to implement `IRequester`).

**Comment by IanStorm** [comment]
The `IRequester` did it. Now all problems are solved.
Thank you for the great support! :)
